lib/stat_tracker.rb

```diff
@@ -87,4 +85,82 @@ def average_goals_by_season
     avg_goals
   end
 
+  def goals_for_average(team_id, filter = nil)
+    goals, games = 0, 0
+    if !filter
+      @game_teams_data.each do |game|
+        goals += game.goals if game.team_id == team_id
+        games += 1 if game.team_id == team_id
+      end
+    else
+      @game_teams_data.each do |game|
+        goals += game.goals if game.team_id == team_id && game.hoa == filter
+        games += 1 if game.team_id == team_id && game.hoa == filter
+      end
+    end
+    (goals.to_f / games).round(2)
+  end
+
+  def goals_against_average(team_id)
+    goals, games = 0, 0
+    @games_data.each do |game|
+      if game.home_team_id == team_id
+        goals += game.away_goals
+        games += 1
+      elsif game.away_team_id == team_id
+        goals += game.home_goals
+        games += 1
+      end
+    end
+    (goals.to_f / games).round(2)
+  end
+
+  def best_offense
+    @teams_data.max_by do |team|
+      goals_against_average(team.team_id)
+    end.teamname
+  end
+
+  def worst_offense
+    @teams_data.min_by do |team|
+      goals_for_average(team.team_id)
+    end.teamname
+  end
+
+  def best_defense
+    @teams_data.min_by do |team|
+      goals_for_average(team.team_id)
+    end.teamname
+  end
+
+  def worst_defense
+    @teams_data.max_by do |team|
+      goals_for_average(team.team_id)
+    end.teamname
+  end
+
+  def highest_scoring_visitor
+    @teams_data.max_by do |team|
+      goals_for_average(team.team_id, "away")
+    end.teamname
+  end
+
+  def highest_scoring_home_team
+    @teams_data.max_by do |team|
+      goals_for_average(team.team_id, "home")
+    end.teamname
+  end
+
+  def lowest_scoring_visitor
+    @teams_data.min_by do |team|
+      goals_for_average(team.team_id, "away")
+    end.teamname
+  end
+
+  def lowest_scoring_home_team
+    @teams_data.min_by do |team|
+      goals_for_average(team.team_id, "home")
+    end.teamname
+  end
+
```

Comment:
Worth noting that we're at 164 lines of code in Stat Tracker mid-way through Iteration 3. Rubric indicates all methods should be < 150 lines... so some further refactoring via inheritance/modules will be quite necessary.